### PR TITLE
refactor: switch to use high_level_rs apis

### PR DESCRIPTION
**Pull request by kazuk**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -15,6 +15,6 @@ name = "springql_client"
 cbindgen = "0.23"
 
 [dependencies]
-springql-core = "0.9.0"
+springql-core = { git="https://github.com/SpringQL/SpringQL.git", branch="refactor_api_2"}
```
Comment (kazuk) [inline Cargo.toml:18]: Change this to
```suggestion
springql-core = "0.11.0"
```

File: src/conf.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    pub fn drop(ptr: *mut SpringConfig) {
+        let outer = unsafe { Box::from_raw(ptr) };
+        let inner = unsafe { Box::from_raw(outer.0) };
+        drop(inner);
+        drop(outer);
+    }
+
+    pub fn into_ptr(self) -> *mut SpringConfig {
+        Box::into_raw(Box::new(self))
+    }
```
Comment (laysakura) [inline src/conf.rs:26]: Nice helper functions. But these functions should not be `pub`.

File: src/conf.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+use ::springql_core::api::SpringConfig as RawSpringConfig;
```
Comment (laysakura) [inline src/conf.rs:1]: You need to add the following header line to all the new files.

```rust
// This file is part of https://github.com/SpringQL/SpringQL-client-c which is licensed under MIT OR Apache-2.0. See file LICENSE-MIT or LICENSE-APACHE for full license details.
```

See: https://github.com/SpringQL/SpringQL-client-c/runs/6767648263?check_suite_focus=true

File: src/pipe_ptr.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+#[non_exhaustive]
+#[repr(transparent)]
+pub struct SpringPipeline(*mut c_void);
+
+impl SpringPipeline {
+    pub fn new(pipe: Pipeline) -> Self {
+        SpringPipeline(unsafe { mem::transmute(Box::new(pipe)) })
+    }
+
+    pub fn pipe(&self) -> &Pipeline {
```
Comment (laysakura) [inline src/pipe_ptr.rs:14]: [nits] like `as_pipeline()` better.

File: src/conf.rs (hunk 3 above)
Comment (laysakura) [reply src/conf.rs:1]: File name: How about renaming into `spring_config.rs`?

File: src/pipe_ptr.rs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+use ::springql_core::api::SpringPipeline as Pipeline;
```
Comment (laysakura) [inline src/pipe_ptr.rs:1]: File name: How about renaming into `spring_pipeline.rs`?

File: src/row_ptr.rs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+use ::springql_core::api::SpringRow as Row;
```
Comment (laysakura) [inline src/row_ptr.rs:1]: File name: How about renaming into `spring_row.rs`?

File: src/row_ptr.rs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+#[non_exhaustive]
+#[repr(transparent)]
+pub struct SpringRow(*mut c_void);
+
+impl SpringRow {
+    pub fn new(pipe: Row) -> Self {
+        SpringRow(unsafe { mem::transmute(Box::new(pipe)) })
+    }
+
+    pub fn row(&self) -> &Row {
```
Comment (laysakura) [inline src/row_ptr.rs:14]: [nits] like `as_row()` better.

File: src/lib.rs [hunk 8]
```diff
@@ -178,13 +164,16 @@ pub unsafe extern "C" fn spring_pop(
...
-
-    with_catch(|| springql_core::spring_pop(pipeline, &queue)).map_or_else(
-        |_| ptr::null_mut(),
-        |row| Box::into_raw(Box::new(SpringRow(mem::transmute(Box::new(row))))),
-    )
+    let result = with_catch(|| pipe.pop(&queue));
+    match result {
+        Ok(row) => {
+            let ptr = SpringRow::new(row);
+            ptr.into_ptr()
```
Comment (laysakura) [inline src/lib.rs:173]:
```suggestion
            let row = SpringRow::new(row);
            row.into_ptr()
```

File: src/lib.rs [hunk 9]
```diff
@@ -178,13 +164,16 @@ pub unsafe extern "C" fn spring_pop(
     pipeline: *const SpringPipeline,
     queue: *const c_char,
 ) -> *mut SpringRow {
-    let pipeline = &*((*pipeline).0 as *const springql_core::SpringPipeline);
+    let pipe = (*pipeline).pipe();
```
Comment (laysakura) [inline src/lib.rs:167]: [nits] I like `let pipeline = ...` better.

File: src/lib.rs [hunk 10]
```diff
@@ -203,24 +192,23 @@ pub unsafe extern "C" fn spring_pop_non_blocking(
...
     let queue = CStr::from_ptr(queue).to_string_lossy().into_owned();
-
-    with_catch(|| springql_core::spring_pop_non_blocking(pipeline, &queue)).map_or_else(
-        |_| {
+    let result = with_catch(|| pipe.pop_non_blocking(&queue));
+    match result {
+        Ok(Some(row)) => {
+            let ptr = SpringRow::new(row);
+            ptr.into_ptr()
```
Comment (laysakura) [inline src/lib.rs:201]: Should set `*is_err = false` in this path.

I recommend to initialize it like https://github.com/SpringQL/SpringQL-client-c/pull/36/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759L206

File: src/lib.rs [hunk 11]
```diff
@@ -355,21 +349,27 @@ pub unsafe extern "C" fn spring_column_text(
...
     let i_col = i_col as usize;
-
-    with_catch(|| springql_core::spring_column_text(row, i_col))
-        .map_or_else(|errno| errno as c_int, |text| strcpy(&text, out, out_len))
+    let result: Result<String, SpringErrno> =
+        with_catch(|| row.get_not_null_by_index(i_col as usize));
+    match result {
+        Ok(v) => {
+            strcpy(&v, out, out_len);
+            v.len() as c_int
```
Comment (laysakura) [inline src/lib.rs:359]: [IMO] Better not to discard the retval from `strcpy()`.

```suggestion
            let v_len = strcpy(&v, out, out_len);
            v_len as c_int
```

**Review by laysakura — CHANGES_REQUESTED** [review]
Thanks. FFI codes get cleaner 😄 

Some comments, mostly related to namings.

File: src/conf.rs (hunk 2 above)
Comment (kazuk) [reply src/conf.rs:26]: No, this struct defined in private `mod`. anything to public to outside lib.

**Comment by kazuk** [comment]
test notify

File: src/conf.rs (hunk 2 above)
Comment (laysakura) [reply src/conf.rs]: Got it.

**Review by laysakura — APPROVED** [review]
👍

**Comment by kazuk** [comment]
thanks review approval.

I commit [this](https://github.com/SpringQL/SpringQL-client-c/pull/36#discussion_r890759800) now.

**Pull request merged**